packages/customize-widgets/src/filters/move-to-sidebar.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { without } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { BlockControls } from '@wordpress/block-editor';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { addFilter } from '@wordpress/hooks';
+
+const { wp } = window;
+
+/**
+ * Internal dependencies
+ */
+import MoveToSidebar from '../components/move-to-sidebar';
+
+/**
+ * Use the customize API to get the setting for the sidebar.
+ * This returns a function that acts as a setter/getter in one for widget ids
+ * in sidebars.
+ *
+ * @param {string} id The id of the sidebar
+ *
+ * @return {Function} A function that when called with no arguments, returns
+ *                    an array of widgetIds for the sidebar. When called with
+ *                    an array, sets that array as the widget ids for the
+ *                    sidebar.
+ */
+function getSidebarWidgetSetting( id ) {
+	return wp.customize( `sidebars_widgets[${ id }]` );
+}
+
+/**
+ * Returns the sidebar object that the widget represented by the provided
+ * widgetId belongs to.
+ *
+ * @param {string} widgetId A widget id
+ *
+ * @return {Object} Configuration of the sidebar that the widget is in.
+ */
+function getSidebarForWidgetId( widgetId ) {
+	return wp.customize.Widgets.data.registeredSidebars.find( ( { id } ) => {
+		const sidebarWidgetsSetting = getSidebarWidgetSetting( id );
+		const sidebarWidgetIds = sidebarWidgetsSetting();
+		return sidebarWidgetIds.includes( widgetId );
+	} );
+}
+
+/**
+ * Returns a list of the registered sidebars.
+ *
+ * @return {Object[]} An array of sidebar configurations objects.
+ */
+function getSidebars() {
+	return wp.customize.Widgets.data.registeredSidebars;
+}
+
+/**
+ * Moves a widget from one sidebar to another.
+ *
+ * @param {string} widgetId     The id of the widget to move.
+ * @param {string} oldSidebarId The id of the sidebar to move from.
+ * @param {string} newSidebarId The id of the sidebar to move to.
+ */
+function moveToSidebar( widgetId, oldSidebarId, newSidebarId ) {
```

Comment:
I think maybe we should implement this method in `sidebar-adaptor`? That's where most of the _editing_ features are being implemented. We could use a context provider to pass down the `sidebar` instance. Not sure it's simpler though.

Reply:
The thing about `SidebarAdapter` though is that it encapsulates actions that can happen on a **single** sidebar. `moveToSidebar` modifies two sidebars.

Still, I agree it'd be nice if we could move `moveToSidebar` so that it's nearer to where all of the widget ↔ block data munging happens. Maybe just something like `import { SidebarAdapter, moveToSidebar } from './munging';` would be OK. (But please don't actually call it `munging` lol.)

Reply:
Huh, or maybe do! 🤷‍♂️

<img width="759" alt="Screen Shot 2021-04-19 at 13 52 33" src="https://user-images.githubusercontent.com/612155/115179352-7e955380-a116-11eb-9097-9ba56933ca20.png">

Reply:
Are you thinking of something like a `customizeAPIProvider` for all interactions with the external API?

Would we want to expose everything from SidebarAdaptor

Are there other functions in the sidebar

Reply:
> The thing about SidebarAdapter though is that it encapsulates actions that can happen on a single sidebar. moveToSidebar modifies two sidebars.

> Still, I agree it'd be nice if we could move moveToSidebar so that it's nearer to where all of the widget ↔ block data munging happens. Maybe just something like import { SidebarAdapter, moveToSidebar } from './munging'; would be OK. (But please don't actually call it munging lol.)

My current interpretation is that `SidebarAdapter` and `moveToSidebar` don't belong together.

If we had a Venn diagram, the customize API would be a bubble on the left side, the block editor API a bubble on the right side, and the Sidebar adapter would be part of the overlapping section that coordinates the two.

I think it might be sensible to have things logically split into three parts in that way.

`moveToSidebar` would be in the left customize API bubble as it purely calls the customize API and nothing else.

Having said that, it still feels a little early to engineer this stuff in this way.

Reply:
In my mind, the `SidebarAdapter` is more like an abstraction of the actual sidebar instance, but not solely only for communicating between block editor and the customize API. We can rename it, or create a new abstraction on top on it. Either way, I think we should encapsulate all the actions we use to manipulate a sidebar in an abstraction rather than calling the customize API in our React code. It just seems easier to me to reason about if all the logics are in the same place and be re-exported to be more _modern_ and semantic in the context of React.

Instead of operating on ids, I would imagine operating on the instance of the abstraction would be better. Hence the `moveToSidebar` function could be rewritten to something like:

```js
Sidebar.prototype.moveWidget = function moveWidget( widget, toSidebar ) {
  this.removeWidget( widget );

  toSidebar.addWidget( widget );
};
```

In theory, we could create abstraction instances for all three `widget`, `sidebar` and `sidebars`, but that feels like a premature optimization for now.

I'm okay with this current approach, we might want to revisit it and refactor or move things around a little bit in the future, but let's not overthink it for now 😅.

Reply:
> It just seems easier to me to reason about if all the logics are in the same place and be re-exported to be more modern and semantic in the context of React.

I see what you mean, I agree that would be nice goal, but as you say I don't think it should be formalized too quickly. As long as these are internal APIs that are easy to refactor, time is on our side to find the right abstraction.

The abstractions you propose would be good, but I think we should avoid coupling this to the adapter part of the application. The adapter could use the abstractions, but not the other way around. I think that's close to what we have now, though I wonder if something like `undo` or `redo` should be part of a sidebar.

In my head, the object oriented nature of it feels a bit unnecessary. There might be reasons for this that I don't know about.